plugin-core/src/main/java/appland/webviews/appMap/AppMapFileEditor.java

```diff
@@ -0,0 +1,235 @@
+                    var location = message.getAsJsonObject("location");
+                    if (location != null) {
+                        return showSource(location.getAsJsonPrimitive("location").getAsString());
+                    }
+                }
+                return null;
+
+            default:
+                return null;
+        }
+    }
+
+    /**
+     * Posts a message to the webview application with new AppMap content to load.
+     */
+    private void reloadAppMapData() {
+        ApplicationManager.getApplication().executeOnPooledThread(() -> {
+            var message = createMessageObject("loadAppMap");
+            message.addProperty("data", createInitData());
+            postMessage(message);
+        });
+    }
+
+    private void setupVfsListener(@NotNull VirtualFile file) {
+        file.getFileSystem().addVirtualFileListener(new VirtualFileListener() {
+            /** This method is called on the EDT */
+            @Override
+            public void contentsChanged(@NotNull VirtualFileEvent event) {
+                assert ApplicationManager.getApplication().isDispatchThread();
+                if (!file.equals(event.getFile())) {
+                    return;
+                }
+
+                LOG.debug("AppMap file was modified, fromSave: " + event.isFromSave() + ", fromRefresh: " + event.isFromRefresh());
+                if (isSelected.get()) {
+                    // load immediately if focused
+                    LOG.debug("contentsChanged: refreshing immediately");
+                    reloadAppMapData();
+                } else {
+                    // delay if not focused
+                    LOG.debug("contentsChanged: delaying refresh");
+                    isModified.set(true);
+                }
+            }
+        }, this);
+    }
+
+    /**
+     * Load the current file's data into the AppLand JS application.
+     */
+    private void applyEditorState(@NotNull AppMapFileEditorState state) {
+        var message = createMessageObject("setAppMapState");
+        message.addProperty("state", state.jsonState);
+        postMessage(message);
+    }
+
+    @Nullable
+    private JBCefJSQuery.Response uploadAppMap() {
+        ApplicationManager.getApplication().invokeLater(() -> {
+            AppMapUploader.uploadAppMap(project, file, url -> {
+                ApplicationManager.getApplication().invokeLater(() -> {
+                    BrowserUtil.browse(url);
+                });
+            });
+        }, ModalityState.defaultModalityState());
+        return null;
+    }
+
+    @Nullable
+    private JBCefJSQuery.Response showSource(String relativePath) {
+        ApplicationManager.getApplication().invokeLater(() -> {
+            var location = FileLocation.parse(relativePath);
+            if (location != null) {
+                var referencedFile = FileLookup.findRelativeFile(project, file, FileUtil.toSystemIndependentName(location.filePath));
```

Comment:
`FileLookup.findRelativeFile` looks unrelated to this change set, but I figure it's worth sharing this. When viewing source:

```
Exception in plugin AppMap (0.23.0).

java.lang.Throwable: Slow operations are prohibited on EDT. See SlowOperations.assertSlowOperationsAreAllowed javadoc.
	at com.intellij.openapi.diagnostic.Logger.error(Logger.java:182)
	at com.intellij.util.SlowOperations.assertSlowOperationsAreAllowed(SlowOperations.java:102)
	at com.intellij.util.indexing.FileBasedIndexImpl.ensureUpToDate(FileBasedIndexImpl.java:784)
	at com.intellij.util.indexing.FileBasedIndexEx.processExceptions(FileBasedIndexEx.java:244)
	at com.intellij.util.indexing.FileBasedIndexEx.processValueIterator(FileBasedIndexEx.java:349)
	at com.intellij.util.indexing.FileBasedIndexEx.processValuesInScope(FileBasedIndexEx.java:318)
	at com.intellij.util.indexing.FileBasedIndexEx.processValues(FileBasedIndexEx.java:223)
	at com.intellij.psi.search.FilenameIndex.getVirtualFilesByName(FilenameIndex.java:210)
	at com.intellij.psi.search.FilenameIndex.getVirtualFilesByName(FilenameIndex.java:57)
	at com.intellij.psi.search.FilenameIndex.getVirtualFilesByName(FilenameIndex.java:52)
	at com.intellij.psi.search.FilenameIndex.getVirtualFilesByName(FilenameIndex.java:65)
	at appland.files.FileLookup.findRelativeFile(FileLookup.java:53)
	at appland.webviews.appMap.AppMapFileEditor.lambda$showSource$5(AppMapFileEditor.java:219)
	at com.intellij.openapi.application.TransactionGuardImpl.runWithWritingAllowed(TransactionGuardImpl.java:214)
	at com.intellij.openapi.application.TransactionGuardImpl.access$200(TransactionGuardImpl.java:21)
	at com.intellij.openapi.application.TransactionGuardImpl$2.run(TransactionGuardImpl.java:196)
	at com.intellij.openapi.application.impl.ApplicationImpl.runIntendedWriteActionOnCurrentThread(ApplicationImpl.java:805)
	at com.intellij.openapi.application.impl.ApplicationImpl.lambda$invokeLater$4(ApplicationImpl.java:348)
	at com.intellij.openapi.application.impl.FlushQueue.doRun(FlushQueue.java:82)
	at com.intellij.openapi.application.impl.FlushQueue.runNextEvent(FlushQueue.java:131)
	at com.intellij.openapi.application.impl.FlushQueue.flushNow(FlushQueue.java:47)
	at com.intellij.openapi.application.impl.FlushQueue$FlushNow.run(FlushQueue.java:187)
	at java.desktop/java.awt.event.InvocationEvent.dispatch(InvocationEvent.java:313)
	at java.desktop/java.awt.EventQueue.dispatchEventImpl(EventQueue.java:776)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:727)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:721)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:85)
	at java.desktop/java.awt.EventQueue.dispatchEvent(EventQueue.java:746)
	at com.intellij.ide.IdeEventQueue.defaultDispatchEvent(IdeEventQueue.java:891)
	at com.intellij.ide.IdeEventQueue._dispatchEvent(IdeEventQueue.java:760)
	at com.intellij.ide.IdeEventQueue.lambda$dispatchEvent$6(IdeEventQueue.java:447)
	at com.intellij.openapi.progress.impl.CoreProgressManager.computePrioritized(CoreProgressManager.java:818)
	at com.intellij.ide.IdeEventQueue.lambda$dispatchEvent$7(IdeEventQueue.java:446)
	at com.intellij.openapi.application.impl.ApplicationImpl.runIntendedWriteActionOnCurrentThread(ApplicationImpl.java:805)
	at com.intellij.ide.IdeEventQueue.dispatchEvent(IdeEventQueue.java:492)
	at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:203)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:124)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:113)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:109)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.desktop/java.awt.EventDispatchThread.run(EventDispatchThread.java:90)
```

Reply:
Nevermind, I see this is resolved in the next pull request.